### PR TITLE
adapter,storage: update controller to support work on source errors

### PR DESCRIPTION
**Pull request by andrioni**
This is the first part of the work on exposing source errors, and it includes the necessary changes in the communication between the coordinator and the storage controller to support the rest of the work.

There are two changes being made here:
- Adding `status_collection_id` as an optional field in `CollectionDescription`, as a way to allow the Adapter to pass an existing `GlobalId` to be used as the collection for storing source status updates.
- Adding `status_shard` as a new field `CollectionMetadata`. It is going to store the ShardId of the collection used to persist errors and other status information for each source. In the initial proposal, a single shared collection will be used for all sources, but having this exposed in `CollectionMetadata` (similarly to `remap_shard` for reclocking) allows us to change this in the future without breaking the stored state for the storage controller (and it makes the code cleaner overall).

I wanted to get this change merged first, as the rest of work should not involve any breaking changes and can be done/reviewed in smaller pieces at a calmer pace, in particular on the Healthchecker.

For more context on this change, please feel free to check the [design doc on Notion](https://www.notion.so/materialize/Exposing-source-errors-via-SQL-1c83b2070b5a48db967b829b7ccd88dc) or the [WIP branch](https://github.com/andrioni/materialize/tree/andrioni/test-storage-v3) from where the next PRs will come from as it stabilizes.

### Motivation
  * This PR makes progress towards a known-desirable feature: #12864

### Tips for reviewer
For Matt and Joe in particular, please feel free to focus only in the changes in Adapter: the biggest one is adding a new `externally_managed` field to tables to track tables whose storage collections are updated/managed outside of the Adapter. This is used to ensure we don't advance timestamps for those tables (as otherwise coord panics), and to avoid hardcoding it for the table we have in that scenario right now (`mz_source_status_history`, for now).

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/adapter/src/coord.rs [hunk 1]
```diff
@@ -1191,9 +1195,14 @@ impl<S: Append + 'static> Coordinator<S> {
     async fn advance_local_inputs(&mut self, advance_to: mz_repr::Timestamp) {
         self.advance_tables.insert(
             advance_to,
-            self.catalog
-                .entries()
-                .filter_map(|e| if e.is_table() { Some(e.id()) } else { None }),
+            self.catalog.entries().filter_map(|e| {
+                // TODO(andrioni): figure out if there's a way to let this not conflict with updates from storaged
```
Comment (mjibson) [inline src/adapter/src/coord.rs:1199]: I don't understand what this TODO is worried about. What are the possible conflicts?

**Review by mjibson — COMMENTED** [review]
These changes look good, but there might be better solutions that I think are worth brainstorming. We currently have two ways for the system to expose its internal metadata for users: 1) tables that the coordinator inserts into, 2) logs that dataflow workers (compute?) send diffs to. Can this feature fit into one of those? Maybe it can't because it needs to take place out side of both coord and compute, so it's now a third thing which is some storage controlled internal metadata store. If that's the correct way to think about it, I'm not sure that making external tables is the right solution. It feels closest to 2, but storage managed instead of compute managed. Does that sound right? If it's right, can we do something like add a variant to `LogVariant` that's about storage instead of compute?

File: src/adapter/src/coord.rs (hunk 1 above)
Comment (andrioni) [reply src/adapter/src/coord.rs:1199]: This only works if noone else appends to the collection, otherwise it panics, as the upper is usually past what it expected it to be.

Comment (mjibson) [reply src/adapter/src/coord.rs:1199]: This further suggests not making this a table then, as that's already what sources and bulitin logs are.

**Comment by andrioni** [comment]
>Maybe it can't because it needs to take place out side of both coord and compute

Indeed, all the data for this specific table is going to be fully managed on `storaged`, and with multiple writers. I left that part out of this PR, but [here's the current version](https://github.com/andrioni/materialize/blob/andrioni/test-storage-v3/src/storage/src/source/healthcheck.rs). There's also more detail in [the design doc](https://www.notion.so/materialize/Exposing-source-errors-via-SQL-1c83b2070b5a48db967b829b7ccd88dc), if you also want to check approaches we considered, but rejected (including sending messages back to let the coordinator handle inserts).

>It feels closest to 2, but storage managed instead of compute managed.

That is also correct, except that this is not dataflow-related (it keeps track of the health status of each source), doesn't use Timely and uses internal data from each `SourceReader`, which is otherwise lost. We are managing the storage collection directly, including reading and appending data to it (and in the future, removing data).

>If it's right, can we do something like add a variant to LogVariant that's about storage instead of compute?

That's a good question! I'm not really familiar with `LogVariant`, so I don't know how it would work. We decided to expose this special collection as a table because that's what felt the simplest way to expose this storage collection to SQL, as doing it otherwise would require adding special SQL syntax (to integrate it with the catalog) that users should not be able to run. I'm fine creating a new not-really-table thing for it, if we think it's better, I just need some pointers on how to expose it through the SQL layer as if it were exactly like a regular table (or source, I guess).

(that said, this collection has to be first created and assigned a `GlobalId` by the coordinator, because we don't to leak storage `ShardId`s out of Storage: that was one of the issues with https://github.com/MaterializeInc/materialize/pull/13593, which was the previous attempt at this)

**Comment by mjibson** [comment]
> I'm fine creating a new not-really-table thing for it, if we think it's better, I just need some pointers on how to expose it through the SQL layer as if it were exactly like a regular table

This is exactly what `BuiltinLog` things already are, you can see how they present as queryable SQL objects in builtin.rs. It's already quite similar to what you have.

**Comment by andrioni** [comment]
Thanks, just pushed a new commit with a new `CatalogType` (provisionally called `BuiltinStorageCollection`, although I'd love to hear suggestions for a better name), is it more or less something like this you had in mind?

File: src/adapter/src/coord.rs [hunk 2]
```diff
@@ -833,6 +833,11 @@ impl<S: Append + 'static> Coordinator<S> {
                         }
                     }
 
+                    let source_status_collection_id =
```
Comment (jkosh44) [inline src/adapter/src/coord.rs:836]: Moving this outside of the for loop would save some time since we'll always get the same ID.

File: src/adapter/src/coord.rs [hunk 3]
```diff
@@ -1191,9 +1207,14 @@ impl<S: Append + 'static> Coordinator<S> {
     async fn advance_local_inputs(&mut self, advance_to: mz_repr::Timestamp) {
         self.advance_tables.insert(
             advance_to,
-            self.catalog
-                .entries()
-                .filter_map(|e| if e.is_table() { Some(e.id()) } else { None }),
+            self.catalog.entries().filter_map(|e| {
+                // TODO(andrioni): figure out if there's a way to let this not conflict with updates from storaged
```
Comment (jkosh44) [inline src/adapter/src/coord.rs:1211]: This comment is probably no longer needed?

**Review by jkosh44 — APPROVED** [review]
Everything in `src/adapter/` looks good to me. I'm assuming that it was on purpose to give sources `Some` `status_collection_id`, while tables and recorded views are given `None`?

I'm also curious why did you decide to create a new catalog object over creating a new log variant? I don't know enough about logs to determine which is the better approach.

File: src/adapter/src/catalog.rs [hunk 4]
```diff
@@ -1082,6 +1096,7 @@ impl CatalogItem {
             CatalogItem::Func(_) => mz_sql::catalog::CatalogItemType::Func,
             CatalogItem::Secret(_) => mz_sql::catalog::CatalogItemType::Secret,
             CatalogItem::Connection(_) => mz_sql::catalog::CatalogItemType::Connection,
+            CatalogItem::StorageCollection(_) => mz_sql::catalog::CatalogItemType::Source,
```
[the remark this answers is not in the document]
Comment (andrioni) [reply src/adapter/src/catalog.rs:1099]: Good question! I'm just reusing it because the logs also reuse it, but I'm not entirely sure of the implications. From what I could tell from looking at the usages of `CatalogItemType`, all the `Source` behaviors are appropriate, but I might be missing something.

File: src/adapter/src/catalog/builtin.rs [hunk 5]
```diff
@@ -1192,6 +1203,23 @@ pub static MZ_AUDIT_EVENTS: Lazy<BuiltinTable> = Lazy::new(|| BuiltinTable {
         .with_column("occurred_at", ScalarType::TimestampTz.nullable(false)),
 });
 
+pub static MZ_SOURCE_STATUS_HISTORY: Lazy<BuiltinStorageCollection> =
```
[the remark this answers is not in the document]
Comment (andrioni) [reply src/adapter/src/catalog/builtin.rs:1206]: Yes! This is coming in a future PR, since the idea is that `storaged` pods are going to advance time for this collection.

File: src/adapter/src/coord.rs (hunk 2 above)
Comment (andrioni) [reply src/adapter/src/coord.rs]: Nice catch, thanks!

File: src/adapter/src/coord.rs (hunk 3 above)
Comment (andrioni) [reply src/adapter/src/coord.rs]: Indeed!

**Comment by mjibson** [comment]
I concur with Joe that this should be a new log variant instead of a new catalogitem type.

**Comment by andrioni** [comment]
I can look more into this tomorrow, but it seemed to me that adding it as `LogVariant` would be harder and require far more changes in `compute`/`compute-client`. The semantics here are far more `Source`-like than `Log`-like overall: it's a collection managed by `storaged`, it gets created by storage controller (as other sources/tables), it can't have dependencies and its data doesn't come from differential collections, but it's instead managed directly by `storaged`.

File: src/adapter/src/catalog.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (benesch) [reply src/adapter/src/catalog.rs:1099]: Yeah, I think thinking of this as a "built-in source" makes sense. There may yet be some safety issues we're not considering, but surely less than there would be if we called this a "table" or an "index"! And I like not introducing another noun that users need to deal with; this should present as a source in all user-facing contexts.

**Comment by benesch** [comment]
> I can look more into this tomorrow, but it seemed to me that adding it as `LogVariant` would be harder and require far more changes in `compute`/`compute-client`. The semantics here are far more `Source`-like than `Log`-like overall: it's a collection managed by `storaged`, it gets created by storage controller (as other sources/tables), it can't have dependencies and its data doesn't come from differential collections, but it's instead managed directly by `storaged`.

I agree! `LogVariant` is presently a compute-owned thing that has lots of special behavior in the coordinator/catalog around getting a fresh ID per cluster. 

> We currently have two ways for the system to expose its internal metadata for users: 1) tables that the coordinator inserts into, 2) logs that dataflow workers (compute?) send diffs to. Can this feature fit into one of those? **Maybe it can't because it needs to take place out side of both coord and compute, so it's now a third thing which is some storage controlled internal metadata store.**

The emphasis (mine) is exactly right!

> If that's the correct way to think about it, I'm not sure that making external tables is the right solution. It feels closest to 2, but storage managed instead of compute managed. Does that sound right?

Yep!

> If it's right, can we do something like add a variant to LogVariant that's about storage instead of compute?

Despite the conceptual similarity, I think, sadly, practically this would cause more problems than it would solve. There's a bunch of code in adapter that handles each of the builtin logs as things that need to be created per compute instance, e.g.

https://github.com/MaterializeInc/materialize/blob/f6aa2c862ea30a79dc3f2d8eed32e6c2a734e60b/src/adapter/src/catalog.rs#L1618-L1625

and none of that will apply to these builtin storage collections.

----

tl;dr I think what @andrioni has here is probably the most straightforward way to add this!

**Review by mjibson — APPROVED** [review]
